kornia/color/zca.py

```diff
@@ -0,0 +1,243 @@
+
+
+def zca_mean(inp: torch.Tensor, eps: float = 1e-7, biased: bool = False,
+             compute_inv: bool = False) -> Tuple[torch.Tensor, ...]:
+    r"""
+
+    Computes ZCA whitening matrix and mean vector. The output could be used in
+    :class:`~torchvision.transforms.LinearTransformation` or with
+    :method:`~kornia.color.zca_whiten`.
+
+    See :class:`~kornia.color.ZCAWhitening` for details.
+
+
+    args:
+        inp (torch.Tensor) : input data tensor
+        eps (float) : a small number used for numerial stablility.
+        biased (bool): Whether to use the biased estimate of the covariance matrix
+        compute_inv (bool): Whether to return the inverse transform matrix
+
+
+    returns:
+        Tuple[torch.Tensor, ...]:
+        A tuple containing the ZCA matrix and the mean vector, and if compute_inv = True, the
+        inverse whitening matrix is retured as well as the final return value.
+
+    Examples:
+        >>> x = torch.tensor([[0,1],[1,0],[-1,0]], dtype = torch.float32)
+        >>> transform, mean = zca_mean(x) # Returns transformation matrix and data mean
+        >>> transform, mean, transform_inv = zca(x, compute_inv = True) # Returns the inverse transform as well.
+
+    """
+
+    if not torch.is_tensor(inp):
+        raise TypeError("Input type is not a torch.Tensor. Got {}".format(
+            type(inp)))
+
+    N: int = inp.size()[0]
+    num_features: int = reduce(lambda a, b: a * b, inp.size()[1::])
+
+    mean: torch.Tensor = torch.mean(inp, dim=0, keepdim=True)
+    mean = mean.view((num_features,))
+
+    inp_center_flat: torch.Tensor = inp.view((N, num_features)) - mean
+
+    cov = inp_center_flat.t().mm(inp_center_flat)
+
+    if biased:
+        cov = cov / float(N)
+    else:
+        cov = cov / float(N - 1)
+
+    U, S, _ = torch.svd(cov)
+
+    S = S.view(-1, 1)
+    S_inv_root: torch.Tensor = torch.rsqrt(S + eps)
+    T: torch.Tensor = (U).mm(S_inv_root * U.t())
+
+    if compute_inv:
+        T_inv: torch.Tensor = (U).mm(torch.sqrt(S) * U.t())
+        return T, mean, T_inv
+
+    return T, mean
+
+
+def zca_whiten(input: torch.Tensor, mean: torch.Tensor,
+               transform: torch.Tensor) -> torch.Tensor:
+    r"""
+
+    Applies and optionally computes the ZCA whitening transform.
+
+    See :class:`~kornia.color.ZCAWhitening` for details.
+
+    args:
+        input (torch.Tensor): Input data
+        mean (torch.Tensor): A tensor containing the mean of the data. If None, this will be computed.
+        transform (torch.Tensor): A tensor containing the CA transformation matrix. If None, this will be computed.
+
+    returns:
+        (torch.Tensor) : Whiten Input data
+
+    Examples:
+        >>> x = torch.tensor([[0,1],[1,0],[-1,0]], dtype = torch.float32)
+        >>> transform, mean = zca_whiten(x)
+        >>> transform, mean, transform_inv = zca(x, mean, transform)
+
+    """
+
+    if mean is None and transform is None:
+        transform, mean = zca_mean(input)
```

Comment:
I would discourage this. To make the API more consistent a bug free I would assume the mean and transform has to be passed all the time. This is something that the user can easily do.

Reply:
I was following this suggestion https://github.com/kornia/kornia/pull/458#discussion_r392694327 but I agree. Should we change the name from zca_whiten to linear_transformation ([similar to Torchvision](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.LinearTransformation)) since without the zca_mean in the function, functionality of the function is similar to LinearTransformation from Torchvision?

Reply:
I think `normalize_zca` should exist. Probably by just removing mean and transform from signature and always calling `mean_zca` + `linear_transform`. In this way we'll introduce this three new functionalities. As for linear_transform, do they really support batch in torchvision? Just to have a clear idea of what would be the difference between functionalities. I see they target this for specific use case, in our case we should think to make it more generic to be used for others use cases. Besides, this last one should end up in kornia.geometry.linalg. Last, check this linear operations group functions to see if we can get any optimization for this https://pytorch.org/docs/stable/torch.html?highlight=bmm#torch.addbmm

Reply:
I agree ```normalize_zca``` should exist and maybe it should be bundle ```mean_zca+linear transform```. Torchvision's [linear transform](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.LinearTransformation) doesn't handle batch since the use case focuses on transforms in dataloaders. Agreed a linear transform function should end up in in kornia.geometry.linalg . I'll take a look at ```torch.addbmm``` seems extremely useful.

Reply:
cool. In that case makes sense to support batched version.